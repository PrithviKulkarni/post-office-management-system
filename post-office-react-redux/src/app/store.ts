import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import letterReducer from '../features/LetterSlice';
import parcelReducer from '../features/ParcelSlice';
import depotReducer from '../features/DepotSlice';
import driverReducer from '../features/DriverSlice' 
export const store = configureStore({
  reducer: {
    letter: letterReducer,
    parcel: parcelReducer,
    depot: depotReducer,
    driver: driverReducer
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
