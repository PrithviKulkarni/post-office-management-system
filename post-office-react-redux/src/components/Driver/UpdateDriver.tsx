import { useState, SyntheticEvent } from 'react'
import { useSelector } from 'react-redux';
import { useParams, useNavigate } from 'react-router-dom';
import { Container, Button, Form, InputGroup } from 'react-bootstrap';
import { Driver, defaultDriver } from '../interfaces/driverinterface';
import { updateADriver, getDriverById } from '../../features/DriverSlice';
import { useAppDispatch } from '../../app/hooks';
import { RootState } from '../../app/store';

const UpdateDriver = () => {
    const navigate = useNavigate();
    const dispatch = useAppDispatch();
    const params = useParams();
    const Driver = useSelector((state: RootState) => getDriverById(state, params.driverId!));

    // const [fname, setFirstname] = useState<string  | undefined>(Driver?.first_name);
    // const [lname, setLastname] = useState<string  | undefined>(Driver?.last_name);
    // const [email, setEmail] = useState<string  | undefined>(Driver?.email);
    // const [fname, setFirstname] = useState<string | undefined>(Driver?.first_name);
    // const [lname, setLastname] = useState<string | undefined>(Driver?.last_name);
    // const [email, setEmail] = useState<string | undefined>(Driver?.email);
    const [fname, setFirstname] = useState<string | undefined>(Driver?.first_name);
    const [lname, setLastname] = useState<string | undefined>(Driver?.last_name);
    const [email, setEmail] = useState<string | undefined>(Driver?.email);
    const [fnameVal, setFirstnameVal] = useState<boolean>(false);
    const [lnameVal, setLastnameVal] = useState<boolean>(false);
    const [emailVal, setEmailVal] = useState<boolean>(false);
    const [errorMessage, setErrorMessage] = useState<string>('')

    /** 
* *
* handles the submit for confirmation and sends a post or patch request to the mongoDB.
* 
*  @returns patch request Drivers. */

    const submitHandler = (event: SyntheticEvent) => {
        event.preventDefault();

        const validated = [fname, lname, email].every(Boolean);
        try {

            if (!fname) {
                setFirstnameVal(true);
            } else {
                setFirstnameVal(false);
            }
            if (!lname) {
                setLastnameVal(true);
            } else {
                setLastnameVal(false);
            }
            if (!email) {
                setEmailVal(true);
            } else {
                setEmailVal(false);
            }

            if (validated) {

                const updDriver: Driver = defaultDriver;
                updDriver.id = Driver!.id;

                if (fname) { updDriver.first_name = fname; }
                if (lname) { updDriver.last_name = lname; }
                if (email) { updDriver.email = email; }

                console.log({ updDriver });

                dispatch(updateADriver(updDriver));
                navigate('/drivers');
            }
        }
        catch (error) {
            setErrorMessage("Driver NOT FOUND.");
            console.log('errorMessage: ', errorMessage);
        }

    }

    return (
        <div className="update-driver" data-testid="update-driver-test" >
            <Container>

                <Form onSubmit={submitHandler}
                    style={{
                        height: "100%",
                        display: "flex",
                        flexDirection: "column",
                        justifyContent: "center",
                        alignItems: "center",
                        width: "100%",
                    }}>
                    <InputGroup>
                        <InputGroup.Text id="basic-addon1">First name</InputGroup.Text>
                        <Form.Control
                            type="text"
                            className="fname"
                            data-testid="update-driver-fname"
                            aria-label="fname"
                            aria-describedby="basic-addon1"
                            size="sm"
                            name="fname"
                            value={fname}
                            onChange={(event) => setFirstname(event.target.value)}
                        />
                        {fnameVal && (
                            <Form.Text className="text-muted" data-testid="firstNameVal">
                                Please fill in field
                            </Form.Text>
                        )}
                    </InputGroup>

                    <InputGroup>
                        <InputGroup.Text id="basic-addon1">Last name</InputGroup.Text>
                        <Form.Control
                            type="text"
                            className="lname"
                            data-testid="update-driver-lname"
                            aria-label="lname"
                            aria-describedby="basic-addon1"
                            size="sm"
                            name="lname"
                            value={lname}
                            onChange={(event) => setLastname(event.target.value)}
                        />
                        {lnameVal && (
                            <Form.Text className="text-muted" data-testid="lastNameVal">
                                Please fill in field
                            </Form.Text>
                        )}
                    </InputGroup>

                    <InputGroup>
                        <InputGroup.Text id="basic-addon1">Email</InputGroup.Text>
                        <Form.Control
                            type="text"
                            aria-label="email"
                            className="email"
                            data-testid="update-driver-email"
                            aria-describedby="basic-addon1"
                            size="sm"
                            name="email"
                            value={email}
                            onChange={(event) => setEmail(event.target.value)}
                        />
                        {emailVal && (
                            <Form.Text className="text-muted" data-testid="emailVal">
                                Please fill in field
                            </Form.Text>
                        )}
                    </InputGroup>

                    <Button className="btn btn-outline-light btn-sm bg-primary" type="submit" data-testid="driver-update-button">Update Driver Details</Button>
                    <p>{errorMessage}</p>
                </Form>
            </Container>
        </div>
    )
}


export default UpdateDriver
