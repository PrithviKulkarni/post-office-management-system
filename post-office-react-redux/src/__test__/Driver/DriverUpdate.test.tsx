import { render, cleanup, waitFor, fireEvent } from '@testing-library/react';
import user from '@testing-library/user-event'
import { screen } from '@testing-library/react';
import { Provider } from "react-redux";
import { store } from '../../app/store';
import { BrowserRouter } from "react-router-dom";
import UpdateDriver from '../../components/Driver/UpdateDriver'

afterEach(cleanup);

describe('Update Form', () => {
     const onSubmit = jest.fn();

     beforeEach(() => {
          onSubmit.mockClear();
     });

     // UPDATE DRIVERS COMPONENT

     it('Renders update Drivers Page Correctly', () => {
          const { asFragment } = render(
               <Provider store={store}>
                    <BrowserRouter>
                         <UpdateDriver />
                    </BrowserRouter>
               </Provider>
          );
          expect(asFragment()).toMatchSnapshot();
     });

     it('Should Verify Update Driver Render Validation', () => {
          const { getByTestId } = render(
               <Provider store={store}>
                    <BrowserRouter>
                         <UpdateDriver />
                    </BrowserRouter>
               </Provider>
          );
          const button = getByTestId("driver-update-button");
          expect(button.textContent).toBe("Update Driver Details");
          fireEvent.click(button);
          expect(button).toBeTruthy();
     
          const firstName = getByTestId("firstNameVal");
          expect(firstName.textContent).toBe("Please fill in field");
          const lastName = getByTestId("lastNameVal");
          expect(lastName.textContent).toBe("Please fill in field");
          const email = getByTestId("emailVal");
          expect(email.textContent).toBe("Please fill in field");
          expect(screen.getByTestId("update-driver-test")).toHaveClass("update-driver");
     
     });
     
     it('Should Verify Update Driver Render When Fields Are Filled', () => {
          const { getByTestId } = render(
               <Provider store={store}>
                    <BrowserRouter>
                         <UpdateDriver />
                    </BrowserRouter>
               </Provider>
          );
          const button = getByTestId("driver-update-button");
          expect(button.textContent).toBe("Update Driver Details");
          const firstName = getByTestId("update-driver-fname");
          fireEvent.change(firstName, { target: { value: "Zainab" } });
          const lastName = getByTestId("update-driver-lname");
          fireEvent.change(lastName, { target: { value: "Bikar" } });
          const email = getByTestId("update-driver-email");
          fireEvent.change(email, { target: { value: "Zainab@gmail.com" } });
          fireEvent.click(button);
          expect(screen.getByTestId("update-driver-test")).toHaveClass("update-driver");
     });
})