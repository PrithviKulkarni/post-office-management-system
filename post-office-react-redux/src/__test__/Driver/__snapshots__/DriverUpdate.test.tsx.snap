// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Form Renders update Drivers Page Correctly 1`] = `
<DocumentFragment>
  <div
    class="update-driver"
    data-testid="update-driver-test"
  >
    <div
      class="container"
    >
      <form
        class=""
        style="height: 100%; display: flex; flex-direction: column; justify-content: center; align-items: center; width: 100%;"
      >
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            First name
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="fname"
            class="fname form-control form-control-sm"
            data-testid="update-driver-fname"
            name="fname"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Last name
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="lname"
            class="lname form-control form-control-sm"
            data-testid="update-driver-lname"
            name="lname"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-group"
        >
          <span
            class="input-group-text"
            id="basic-addon1"
          >
            Email
          </span>
          <input
            aria-describedby="basic-addon1"
            aria-label="email"
            class="email form-control form-control-sm"
            data-testid="update-driver-email"
            name="email"
            type="text"
            value=""
          />
        </div>
        <button
          class="btn btn-outline-light btn-sm bg-primary btn btn-primary"
          data-testid="driver-update-button"
          type="submit"
        >
          Update Driver Details
        </button>
        <p />
      </form>
    </div>
  </div>
</DocumentFragment>
`;
